### ISSUE TITLE
range .until evaluates .front twice

### ISSUE THREAD

**Issue by dlangBugzillaToGithub** [dlang/phobos#9864]
### Description

````markdown
```
import std.algorithm : map, until;
import std.range : iota;
import std.array : array;

int access;
int logAccess(int x) {
    access++;
    return x;
}

void main()
{
    iota(10).map!logAccess.until(10).array;
    
    assert(access == 20);
}
```

if you remove the `.until(10)`, you get `access == 10`
````

**Comment by dlangBugzillaToGithub** [comment]
#### d.bugs commented on 2023-09-20T01:46:17Z



````markdown
`std.algorithm:cache` could be used, but that still evalutes .front one time too many compared to when `until` should stop when No.openRight is passed in as argument, e.g.

```
iota(10).map!logAccess.cache.until(5, No.openRight).array;
```

will call logAccess on 0, 1, 2, 3, 4, 5 and 6
````

**Comment by pbackus** [comment]
The additional call to `front` is not a bug in `until`; it is the documented behavior of `cache`:

> `cache` eagerly evaluates `front` of range on each construction **or call to `popFront`**, to store the result in a cache. [Emphasis added.]

Perhaps the correct solution to this is to add a new `lazyCache` range, which only evaluates `front` on demand.

**Comment by jmdavis** [comment]
I would argue that there is no bug here at all. It is perfectly valid for an algorithm to call `front` multiple times on the same element.

Perhaps `lazyCache` would make sense for some use cases where someone wants to cache the result without it being eager, but the fact that a particular algorithm ends up calling `front` multiple times is not a bug. Part of the expected behavior of the range API is that you can call `front` multiple times and get the same result each time.

If a particular algorithm can be better optimized to call `front` fewer times, then that might be desirable, but it's not at all a requirement that an algorithm call `front` only once per element. And whether it's even more efficient to call `front` only once can depend heavily on the code, because in some cases, the alternative is to copy `front` to able to use the value multiple times without calling `front` multiple times, and that could be more or less efficient depending on the code or the types involved. It also doesn't work with non-copyable types, but most of std.algorithm doesn't work with them anyway.

So, honestly, I'm inclined to just close this bug as not being a bug. The potential addition of something like `lazyCache` can be a separate enhancement request.

**Comment by jmdavis** [comment]
As calling `front` multiple times is not a bug, I'm closing this as "won't fix", but I have created https://github.com/dlang/phobos/issues/10687 as an enhancement request for a lazy version of `cache`.
